test/eth.subscribe.ganache.js

```diff
@@ -0,0 +1,99 @@
+const assert = require('assert');
+const ganache = require('ganache-cli');
+const pify = require('pify');
+const Basic = require('./sources/Basic');
+const utils = require('./helpers/test.utils');
+const Web3 = utils.getWeb3();
+
+describe('subscription connect/reconnect', function(){
+    let server;
+    let web3;
+    let accounts;
+    const port = 8545;
+
+    beforeEach(async function(){
+        server = ganache.server({port: port});
+        await pify(server.listen)(port);
+        web3 = new Web3('ws://localhost:' + port);
+        accounts = await web3.eth.getAccounts();
+    })
+
+    afterEach(async function(){
+        // Might already be closed..
+        try {
+            await pify(server.close)()
+        } catch (err) {}
+    })
+
+    it('subscribes (baseline)', function(){
+        return new Promise(async resolve => {
+            web3.eth
+                .subscribe("newBlockHeaders")
+                .on('data', function(result){
+                    assert(result.parentHash);
+                    this.removeAllListeners();
+                    resolve();
+                });
+
+            await utils.mine(web3, accounts[0]);
+        });
+    })
+
+    it('errors when there is no client to connect to (baseline)', async function(){
+        await pify(server.close)();
+
+        //console.log(util.inspect(web3.eth.currentProvider, true, 5, true))
+        return new Promise(resolve => {
+            web3.eth
+                .subscribe("newBlockHeaders")
+                .on('error', function(err){
+                    assert(err.message.includes("CONNECTION ERROR"))
+                    this.removeAllListeners();
+                    resolve();
+                });
+        });
+    });
+
+    // This test failing....
+    it.only('auto reconnects', function(){
```

Comment:
@nivida This is my attempt to model how I imagine reconnection working. 
+ Open server
+ Attach a subscription handler
+ Fire a tx (which subscrip hears)
+ Close server
+ Re-open server
+ Fire a tx (which subscrip hears)

But the test doesn't reconnect...do you want to take a look?

Reply:
We need to make sure this works against an actual client, not just mock some assumptions in.

Reply:
Maybe my model of how this works is wrong...not sure.

Reply:
@nivida mentioned something in a chat that's important which I'm leaving here for further reference. 

The way this test models disconnection doesn't reflect the connectivity issues people typically have online e.g. nodes are not shutting down, killing their sockets and restarting...